File: ios/Inferra/LocalServerVPNManager.m
#import <Foundation/Foundation.h>
#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>
#import <NetworkExtension/NetworkExtension.h>

static NSString *const kLocalServerVPNEvent = @"LocalServerVPNStatusChanged";
static NSString *const kLocalServerVPNBundleId = @"com.gorai.inferra.InferraServerExtension";

@interface LocalServerVPNManager : RCTEventEmitter <RCTBridgeModule>
@end

@interface LocalServerVPNManager ()
@property (nonatomic, strong) NETunnelProviderManager *tunnelManager;
@property (nonatomic, assign) NEVPNStatus cachedStatus;
@end

@implementation LocalServerVPNManager

RCT_EXPORT_MODULE();

- (instancetype)init
{
    self = [super init];
    if (self) {
        _cachedStatus = NEVPNStatusInvalid;
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(handleStatusChange) name:NEVPNStatusDidChangeNotification object:nil];
    }
    return self;
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (NSArray<NSString *> *)supportedEvents
{
    return @[kLocalServerVPNEvent];
}

- (void)startObserving
{
    [self sendStatusEvent];
}

- (void)stopObserving
{
}

+ (BOOL)requiresMainQueueSetup
{
    return YES;
}

RCT_EXPORT_METHOD(startVPNServer:(NSDictionary *)options
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
    [self loadManagerWithCompletion:^(NETunnelProviderManager *manager, NSError *error) {
        if (error) {
            reject(@"vpn_load_failed", error.localizedDescription, error);
            return;
        }
        if (!manager) {
            reject(@"vpn_missing_manager", @"Manager unavailable", nil);
            return;
        }
        NETunnelProviderProtocol *configuration = [NETunnelProviderProtocol new];
        configuration.providerBundleIdentifier = kLocalServerVPNBundleId;
        configuration.serverAddress = @"127.0.0.1";
        configuration.providerConfiguration = [self normalizedConfigurationFromOptions:options];
        manager.protocolConfiguration = configuration;
        manager.localizedDescription = @"Inferra Local Server";
        manager.enabled = YES;
        [manager saveToPreferencesWithCompletionHandler:^(NSError *saveError) {
            if (saveError) {
                reject(@"vpn_save_failed", saveError.localizedDescription, saveError);
                return;
            }
            NSError *startError = nil;
            NSDictionary<NSString *, NSObject *> *startOptions = [self normalizedStartOptionsFromOptions:options];
            BOOL started = [manager.connection startVPNTunnelWithOptions:startOptions andReturnError:&startError];
            if (!started || startError) {
                reject(@"vpn_start_failed", startError.localizedDescription ?: @"Unable to start tunnel", startError);
                return;
            }
            self.tunnelManager = manager;
            self.cachedStatus = manager.connection.status;
            [self sendStatusEvent];
            NSMutableDictionary *payload = [NSMutableDictionary dictionary];
            payload[@"success"] = @YES;
            payload[@"status"] = [self statusStringForStatus:self.cachedStatus];
            id urlValue = options[@"url"];
            if (urlValue) {
                payload[@"url"] = urlValue;
            }
            resolve(payload);
        }];
    }];
}

RCT_EXPORT_METHOD(stopVPNServer:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
    [self loadManagerWithCompletion:^(NETunnelProviderManager *manager, NSError *error) {
        if (error) {
            reject(@"vpn_load_failed", error.localizedDescription, error);
            return;
        }
        if (!manager) {
            resolve(@{ @"success": @NO });
            return;
        }
        [manager.connection stopVPNTunnel];
        self.cachedStatus = manager.connection.status;
        [self sendStatusEvent];
        resolve(@{ @"success": @YES });
    }];
}

RCT_EXPORT_METHOD(getStatus:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
    [self loadManagerWithCompletion:^(NETunnelProviderManager *manager, NSError *error) {
        if (error) {
            reject(@"vpn_load_failed", error.localizedDescription, error);
            return;
        }
        NEVPNStatus status = manager ? manager.connection.status : self.cachedStatus;
        NSDictionary *info = [self statusPayloadForManager:manager status:status];
        resolve(info);
    }];
}

RCT_EXPORT_METHOD(updateVPNConfiguration:(NSDictionary *)options
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
    [self loadManagerWithCompletion:^(NETunnelProviderManager *manager, NSError *error) {
        if (error) {
            reject(@"vpn_load_failed", error.localizedDescription, error);
            return;
        }
        if (!manager) {
            resolve(@{ @"success": @NO });
            return;
        }
        NETunnelProviderProtocol *protocol = (NETunnelProviderProtocol *)manager.protocolConfiguration;
        if (!protocol) {
            protocol = [NETunnelProviderProtocol new];
        }
        protocol.providerBundleIdentifier = kLocalServerVPNBundleId;
        protocol.serverAddress = @"127.0.0.1";
        NSMutableDictionary *providerConfig = [NSMutableDictionary dictionaryWithDictionary:protocol.providerConfiguration ?: @{}];
        id urlValue = options[@"url"];
        if (urlValue) {
            providerConfig[@"url"] = urlValue;
        }
        protocol.providerConfiguration = providerConfig;
        manager.protocolConfiguration = protocol;
        manager.enabled = YES;
        [manager saveToPreferencesWithCompletionHandler:^(NSError *saveError) {
            if (saveError) {
                reject(@"vpn_save_failed", saveError.localizedDescription, saveError);
                return;
            }
            resolve(@{ @"success": @YES });
        }];
    }];
}

RCT_EXPORT_METHOD(requestVPNPermission:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
    [self loadManagerWithCompletion:^(NETunnelProviderManager *manager, NSError *error) {
        if (error) {
            reject(@"vpn_permission_failed", error.localizedDescription, error);
            return;
        }
        if (!manager) {
            reject(@"vpn_permission_failed", @"Manager unavailable", nil);
            return;
        }
        NETunnelProviderProtocol *protocol = (NETunnelProviderProtocol *)manager.protocolConfiguration;
        if (!protocol) {
            protocol = [NETunnelProviderProtocol new];
        }
        protocol.providerBundleIdentifier = kLocalServerVPNBundleId;
        protocol.serverAddress = @"127.0.0.1";
        if (!protocol.providerConfiguration) {
            protocol.providerConfiguration = @{};
        }
        manager.protocolConfiguration = protocol;
        manager.enabled = NO;
        [manager saveToPreferencesWithCompletionHandler:^(NSError *saveError) {
            if (saveError) {
                reject(@"vpn_permission_failed", saveError.localizedDescription, saveError);
                return;
            }
            [manager loadFromPreferencesWithCompletionHandler:^(NSError *loadError) {
                if (loadError) {
                    reject(@"vpn_permission_failed", loadError.localizedDescription, loadError);
                    return;
                }
                self.cachedStatus = manager.connection.status;
                resolve(@{ @"success": @YES, @"status": [self statusStringForStatus:self.cachedStatus] });
            }];
        }];
    }];
}

- (void)loadManagerWithCompletion:(void (^)(NETunnelProviderManager *manager, NSError *error))completion
{
    [NETunnelProviderManager loadAllFromPreferencesWithCompletionHandler:^(NSArray<NETunnelProviderManager *> *managers, NSError *error) {
        if (error) {
            completion(nil, error);
            return;
        }
        NETunnelProviderManager *matched = nil;
        for (NETunnelProviderManager *candidate in managers) {
            NETunnelProviderProtocol *protocol = (NETunnelProviderProtocol *)candidate.protocolConfiguration;
            if (protocol && [protocol.providerBundleIdentifier isEqualToString:kLocalServerVPNBundleId]) {
                matched = candidate;
                break;
            }
        }
        if (!matched) {
            matched = [NETunnelProviderManager new];
        }
        self.tunnelManager = matched;
        self.cachedStatus = matched.connection.status;
        completion(matched, nil);
    }];
}

- (NSDictionary *)normalizedConfigurationFromOptions:(NSDictionary *)options
{
    NSMutableDictionary *config = [NSMutableDictionary dictionary];
    NSNumber *port = options[@"port"];
    if (port) {
        config[@"port"] = port;
    }
    id urlValue = options[@"url"];
    if (urlValue) {
        config[@"url"] = urlValue;
    }
    return config;
}

- (NSDictionary<NSString *, NSObject *> *)normalizedStartOptionsFromOptions:(NSDictionary *)options
{
    NSMutableDictionary<NSString *, NSObject *> *result = [NSMutableDictionary dictionary];
    NSNumber *port = options[@"port"];
    if (port) {
        result[@"port"] = port;
    }
    NSString *urlValue = options[@"url"];
    if (urlValue) {
        result[@"url"] = urlValue;
    }
    return result.count > 0 ? result : nil;
}

- (NSString *)statusStringForStatus:(NEVPNStatus)status
{
    switch (status) {
        case NEVPNStatusConnected:
            return @"connected";
        case NEVPNStatusConnecting:
            return @"connecting";
        case NEVPNStatusDisconnected:
            return @"disconnected";
        case NEVPNStatusDisconnecting:
            return @"disconnecting";
        case NEVPNStatusInvalid:
            return @"invalid";
        case NEVPNStatusReasserting:
            return @"reasserting";
    }
    return @"unknown";
}

- (NSDictionary *)statusPayloadForManager:(NETunnelProviderManager *)manager status:(NEVPNStatus)status
{
    NSMutableDictionary *info = [NSMutableDictionary dictionary];
    info[@"isRunning"] = @((status == NEVPNStatusConnected) || (status == NEVPNStatusConnecting));
    info[@"status"] = [self statusStringForStatus:status];
    NETunnelProviderProtocol *protocol = (NETunnelProviderProtocol *)manager.protocolConfiguration;
    NSString *urlValue = (NSString *)protocol.providerConfiguration[@"url"];
    if (urlValue) {
        info[@"url"] = urlValue;
    }
    return info;
}

- (void)sendStatusEvent
{
    NETunnelProviderManager *manager = self.tunnelManager;
    NEVPNStatus status = manager ? manager.connection.status : self.cachedStatus;
    NSDictionary *payload = [self statusPayloadForManager:manager status:status];
    dispatch_async(dispatch_get_main_queue(), ^{
        [self sendEventWithName:kLocalServerVPNEvent body:payload];
    });
}

- (void)handleStatusChange
{
    NETunnelProviderManager *manager = self.tunnelManager;
    if (manager) {
        self.cachedStatus = manager.connection.status;
    }
    [self sendStatusEvent];
}

@end

File: ios/Inferra/LocalServerVPNManager.swift
import Foundation
import NetworkExtension
import React

@objc(LocalServerVPNManager)
class LocalServerVPNManager: RCTEventEmitter {
    private let providerBundleIdentifier = "com.gorai.inferra.InferraServerExtension"
    private var tunnelManager: NETunnelProviderManager?
    private var cachedStatus: NEVPNStatus = .invalid

    override init() {
        super.init()
        NotificationCenter.default.addObserver(self, selector: #selector(handleStatusChange), name: .NEVPNStatusDidChange, object: nil)
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    override func supportedEvents() -> [String]! {
        return ["LocalServerVPNStatusChanged"]
    }

    override class func requiresMainQueueSetup() -> Bool {
        return true
    }

    @objc override func startObserving() {
        sendStatusEvent()
    }

    @objc override func stopObserving() {
    }

    @objc
    func startVPNServer(_ options: NSDictionary, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        loadManager { [weak self] manager, error in
            guard let self = self else { return }
            if let error = error {
                reject("vpn_load_failed", error.localizedDescription, error)
                return
            }
            guard let manager = manager else {
                reject("vpn_missing_manager", "Manager unavailable", nil)
                return
            }
            let protocolConfiguration = NETunnelProviderProtocol()
            protocolConfiguration.providerBundleIdentifier = self.providerBundleIdentifier
            protocolConfiguration.serverAddress = "127.0.0.1"
            protocolConfiguration.providerConfiguration = self.normalizedConfiguration(from: options)
            manager.protocolConfiguration = protocolConfiguration
            manager.localizedDescription = "Inferra Local Server"
            manager.isEnabled = true
            manager.saveToPreferences { saveError in
                if let saveError = saveError {
                    reject("vpn_save_failed", saveError.localizedDescription, saveError)
                    return
                }
                do {
                    let startOptions = self.normalizedStartOptions(from: options)
                    try manager.connection.startVPNTunnel(options: startOptions)
                    self.tunnelManager = manager
                    self.cachedStatus = manager.connection.status
                    self.sendStatusEvent()
                    var payload: [String: Any] = [
                        "success": true,
                        "status": self.statusString(self.cachedStatus)
                    ]
                    if let urlValue = options["url"] as? String {
                        payload["url"] = urlValue
                    }
                    resolve(payload)
                } catch let startError {
                    reject("vpn_start_failed", startError.localizedDescription, startError)
                }
            }
        }
    }

    @objc
    func stopVPNServer(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        loadManager { [weak self] manager, error in
            guard let self = self else { return }
            if let error = error {
                reject("vpn_load_failed", error.localizedDescription, error)
                return
            }
            guard let manager = manager else {
                resolve(["success": false])
                return
            }
            manager.connection.stopVPNTunnel()
            self.cachedStatus = manager.connection.status
            self.sendStatusEvent()
            resolve(["success": true])
        }
    }

    @objc
    func getStatus(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        loadManager { [weak self] manager, error in
            guard let self = self else { return }
            if let error = error {
                reject("vpn_load_failed", error.localizedDescription, error)
                return
            }
            let status = manager?.connection.status ?? self.cachedStatus
            let configuration = manager?.protocolConfiguration as? NETunnelProviderProtocol
            let info: [String: Any] = [
                "isRunning": status == .connected || status == .connecting,
                "status": self.statusString(status),
                "url": configuration?.providerConfiguration?["url"]
            ]
            resolve(info)
        }
    }

    @objc
    func updateVPNConfiguration(_ options: NSDictionary, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        loadManager { [weak self] manager, error in
            guard let self = self else { return }
            if let error = error {
                reject("vpn_load_failed", error.localizedDescription, error)
                return
            }
            guard let manager = manager else {
                resolve(["success": false])
                return
            }
            let configuration = manager.protocolConfiguration as? NETunnelProviderProtocol ?? NETunnelProviderProtocol()
            configuration.providerBundleIdentifier = self.providerBundleIdentifier
            configuration.serverAddress = "127.0.0.1"
            var providerConfig = configuration.providerConfiguration ?? [:]
            if let url = options["url"] as? String {
                providerConfig["url"] = url
            }
            configuration.providerConfiguration = providerConfig
            manager.protocolConfiguration = configuration
            manager.isEnabled = true
            manager.saveToPreferences { saveError in
                if let saveError = saveError {
                    reject("vpn_save_failed", saveError.localizedDescription, saveError)
                    return
                }
                resolve(["success": true])
            }
        }
    }

    private func normalizedConfiguration(from options: NSDictionary) -> [String: Any] {
        var config: [String: Any] = [:]
        if let port = options["port"] as? Int {
            config["port"] = port
        }
        if let url = options["url"] as? String {
            config["url"] = url
        }
        return config
    }

    private func normalizedStartOptions(from options: NSDictionary) -> [String: NSObject]? {
        var result: [String: NSObject] = [:]
        if let port = options["port"] as? Int {
            result["port"] = NSNumber(value: port)
        }
        if let url = options["url"] as? String {
            result["url"] = url as NSString
        }
        return result.isEmpty ? nil : result
    }

    private func loadManager(completion: @escaping (NETunnelProviderManager?, Error?) -> Void) {
        NETunnelProviderManager.loadAllFromPreferences { managers, error in
            if let error = error {
                completion(nil, error)
                return
            }
            if let existing = managers?.first(where: { manager in
                guard let configuration = manager.protocolConfiguration as? NETunnelProviderProtocol else {
                    return false
                }
                return configuration.providerBundleIdentifier == self.providerBundleIdentifier
            }) {
                self.tunnelManager = existing
                self.cachedStatus = existing.connection.status
                completion(existing, nil)
                return
            }
            let manager = NETunnelProviderManager()
            self.tunnelManager = manager
            self.cachedStatus = manager.connection.status
            completion(manager, nil)
        }
    }

    @objc private func handleStatusChange() {
        if let status = tunnelManager?.connection.status {
            cachedStatus = status
        }
        sendStatusEvent()
    }

    private func statusString(_ status: NEVPNStatus) -> String {
        switch status {
        case .connected:
            return "connected"
        case .connecting:
            return "connecting"
        case .disconnected:
            return "disconnected"
        case .disconnecting:
            return "disconnecting"
        case .invalid:
            return "invalid"
        case .reasserting:
            return "reasserting"
        @unknown default:
            return "unknown"
        }
    }

    private func sendStatusEvent() {
        let status = tunnelManager?.connection.status ?? cachedStatus
        var info: [String: Any] = [
            "isRunning": status == .connected || status == .connecting,
            "status": statusString(status)
        ]
        if let configuration = tunnelManager?.protocolConfiguration as? NETunnelProviderProtocol {
            if let url = configuration.providerConfiguration?["url"] as? String {
                info["url"] = url
            }
        }
        sendEvent(withName: "LocalServerVPNStatusChanged", body: info)
    }
}
